### PR TITLE
Entity indexing shall happen in separate transaction

### PR DESCRIPTION
**Pull request by ko5tik**
Fix https://github.com/jhipster/generator-jhipster/issues/17488

This is updated and rebased version of:

https://github.com/jhipster/generator-jhipster/pull/17780

which  is updated version of:

https://github.com/jhipster/generator-jhipster/pull/17561

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 1]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
 <%_ } else { _%>
 
 interface <%= entityClass %>SearchRepositoryInternal {
   <%_ if (paginationNo) { _%>
-    Stream<<%= persistClass %>> search(String query);
+    Stream<<%= entityClass %>> search(String query);
+    Stream<<%= entityClass %>> search(Query query);
   <%_ } else { _%>
-    Page<<%= persistClass %>> search(String query, Pageable pageable);
+    Page<<%= entityClass %>> search(String query, Pageable pageable);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:117]:
```suggestion
    Page<<%= persistClass %>> search(String query, Pageable pageable);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 2]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
+    Page<<%= entityClass %>> search(Query query);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:118]:
```suggestion
    Page<<%= persistClass %>> search(Query query);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 3]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
   <%_ } _%>
+
+  void index(<%= entityClass %> entity);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:121]:
```suggestion
  void index(<%= persistClass %> entity);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 4]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
   <%_ if (paginationNo) { _%>
     @Override
-    public Stream<<%= persistClass %>> search(String query) {
+    public Stream<<%= entityClass %>> search(String query) {
         NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
+        return search(nativeSearchQuery);
+    }
+
+    @Override
+    public Stream<<%= entityClass %>> search(Query query) {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:142]:
```suggestion
    public Stream<<%= persistClass %>> search(Query query) {
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 5]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
         return elasticsearchTemplate
-            .search(nativeSearchQuery, <%= persistClass %>.class)
+            .search(query, <%= entityClass %>.class)
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:144]:
```suggestion
            .search(query, <%= persistClass %>.class)
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 6]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
             .map(SearchHit::getContent)
             .stream();
     }
   <%_ } else { _%>
     @Override
-    public Page<<%= persistClass %>> search(String query, Pageable pageable) {
+    public Page<<%= entityClass %>> search(String query, Pageable pageable) {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:150]:
```suggestion
    public Page<<%= persistClass %>> search(String query, Pageable pageable) {
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 7]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
         NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
-        nativeSearchQuery.setPageable(pageable);
-        List<<%= persistClass %>> hits = elasticsearchTemplate.search(nativeSearchQuery, <%= persistClass %>.class).map(SearchHit::getContent)
+        return search(nativeSearchQuery.setPageable(pageable));
+    }
+
+    @Override
+    public Page<<%= entityClass %>> search(Query query) {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:156]:
```suggestion
    public Page<<%= persistClass %>> search(Query query) {
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 8]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
+        List<<%= entityClass %>> hits = elasticsearchTemplate.search(query, <%= entityClass %>.class).map(SearchHit::getContent)
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:157]:
```suggestion
        List<<%= persistClass %>> hits = elasticsearchTemplate.search(query, <%= entityClass %>.class).map(SearchHit::getContent)
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 9]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
 
-        return new PageImpl<>(hits, pageable, hits.size());
+        return new PageImpl<>(hits, query.getPageable(), hits.size());
     }
   <%_ } _%>
+
+      @Override
+      @Async
+      @Transactional(propagation =  Propagation.REQUIRES_NEW, readOnly = true, isolation = Isolation.READ_COMMITTED)
+      public void index(<%= entityClass %> entity) {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:168]:
```suggestion
      public void index(<%= persistClass %> entity) {
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 10]
```diff
@@ -72,54 +83,91 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
-    <%= entityClass %>SearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate) {
+    <%= entityClass %>SearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate, <%= entityClass %>Repository repository) {
         this.elasticsearchTemplate = elasticsearchTemplate;
+        this.repository = repository;
     }
 
   <%_ if (paginationNo) { _%>
     @Override
-    public Stream<<%= persistClass %>> search(String query) {
+    public Stream<<%= entityClass %>> search(String query) {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:136]:
```suggestion
    public Stream<<%= persistClass %>> search(String query) {
```

**Review by mshima — COMMENTED** [review]
Lots of fixes

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 11]
```diff
@@ -87,39 +112,62 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
 interface <%= entityClass %>SearchRepositoryInternal {
   <%_ if (paginationNo) { _%>
     Stream<<%= persistClass %>> search(String query);
+    Stream<<%= persistClass %>> search(Query query);
   <%_ } else { _%>
     Page<<%= persistClass %>> search(String query, Pageable pageable);
+    Page<<%= persistClass %>> search(Query query);
   <%_ } _%>
+
+  void index(<%= persistClass %> entity);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:121]:
```suggestion
    void index(<%= persistClass %> entity);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 12]
```diff
@@ -87,39 +112,62 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
 interface <%= entityClass %>SearchRepositoryInternal {
   <%_ if (paginationNo) { _%>
     Stream<<%= persistClass %>> search(String query);
+    Stream<<%= persistClass %>> search(Query query);
   <%_ } else { _%>
     Page<<%= persistClass %>> search(String query, Pageable pageable);
+    Page<<%= persistClass %>> search(Query query);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:118]:
```suggestion

    Page<<%= persistClass %>> search(Query query);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 13]
```diff
@@ -87,39 +112,62 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
 interface <%= entityClass %>SearchRepositoryInternal {
   <%_ if (paginationNo) { _%>
     Stream<<%= persistClass %>> search(String query);
+    Stream<<%= persistClass %>> search(Query query);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:115]:
```suggestion

    Stream<<%= persistClass %>> search(Query query);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 14]
```diff
@@ -87,39 +112,62 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
     public Page<<%= persistClass %>> search(String query, Pageable pageable) {
         NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
-        nativeSearchQuery.setPageable(pageable);
-        List<<%= persistClass %>> hits = elasticsearchTemplate.search(nativeSearchQuery, <%= persistClass %>.class).map(SearchHit::getContent)
+        return search(nativeSearchQuery.setPageable(pageable));
+    }
+
+    @Override
+    public Page<<%= persistClass %>> search(Query query) {
+        List<<%= persistClass %>> hits = elasticsearchTemplate.search(query, <%= entityClass %>.class).map(SearchHit::getContent)
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:157]:
```suggestion
        List<<%= persistClass %>> hits = elasticsearchTemplate.search(query, <%= persistClass %>.class).map(SearchHit::getContent)
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 15]
```diff
@@ -73,10 +84,24 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
+        builders.stream().forEach(builder -> {
+            queryBuilder.withSort(builder);
+        });
+
+        NativeSearchQuery nativeSearchQuery = queryBuilder.build();
+  <%_ } else { _%>
+        NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
+  <%_ } _%>
+        return search(nativeSearchQuery);
+    }
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:102]:
```suggestion
    }

```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 16]
```diff
@@ -61,6 +71,7 @@ public interface <%= entityClass %>SearchRepository extends <% if (reactive) {%>
 
 interface <%= entityClass %>SearchRepositoryInternal {
     Flux<<%= persistClass %>> search(String query<% if (!paginationNo) { %>, Pageable pageable<% } %>);
+    Flux<<%= persistClass %>> search(Query query);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:74]:
```suggestion

    Flux<<%= persistClass %>> search(Query query);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 17]
```diff
@@ -29,6 +30,11 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
   <%_ } _%>
 import org.springframework.data.domain.Pageable;
+import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
+import org.elasticsearch.search.sort.FieldSortBuilder;
+import java.util.List;
+import org.springframework.data.domain.PageRequest;
+
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:37]:
```suggestion
```

**Review by mshima — COMMENTED** [review]
Prettify

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 18]
```diff
@@ -87,39 +113,64 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
-        return new PageImpl<>(hits, pageable, hits.size());
+        return new PageImpl<>(hits, query.getPageable(), hits.size());
     }
   <%_ } _%>
+
+      @Override
+      @Async
+      @Transactional(propagation =  Propagation.REQUIRES_NEW, readOnly = true, isolation = Isolation.READ_COMMITTED)
+      public void index(<%= persistClass %> entity) {
+         repository.findOneWithEagerRelationships(entity.getId()).ifPresent(elasticsearchTemplate::save);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:172]:
```suggestion
         repository.findOne<% if (implementsEagerLoadApis ) { %>WithEagerRelationships<% } %>(entity.getId()).ifPresent(elasticsearchTemplate::save);
```

**Review by mshima — COMMENTED** [review]
Change find api.

File: generators/entity-server/templates/partials/save_template.ejs [hunk 19]
```diff
@@ -46,18 +46,18 @@ if (isUsingMapsId) {
         <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>().get<%= primaryKey.nameCapitalized %>();
         <%= mapsIdRepoInstance %>.findById(<%= otherEntityName %>Id).ifPresent(<%= persistInstance %>::<%_ if (!fluentMethods) { _%>set<%= mapsIdAssoc.relationshipNameCapitalized %> <%_ } else { _%><%= mapsIdAssoc.relationshipName %><%_ } _%>);
     <%_ } _%>
-        <%= persistInstance %> = <%= entityInstance %>Repository.save(<%= persistInstance %>);
+        <%= persistInstance %> = <%= entityInstance %>Repository.saveAndFlush(<%= persistInstance %>);
```
Comment (mshima) [inline generators/entity-server/templates/partials/save_template.ejs:49]:
```suggestion
        <%= persistInstance %> = <%= entityInstance %>Repository.save(<%= persistInstance %>);
```

File: generators/entity-server/templates/partials/save_template.ejs [hunk 20]
```diff
@@ -46,18 +46,18 @@ if (isUsingMapsId) {
...
         <%= returnPrefix %> <%= entityToDto %>(<%= persistInstance %>);
   <%_ } else {
       resultEntity = 'result'; _%>
     <%_ if (isUsingMapsId) { _%>
         <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>().get<%= primaryKey.nameCapitalized %>();
         <%= mapsIdRepoInstance %>.findById(<%= otherEntityName %>Id).ifPresent(<%= instanceName %>::<%_ if (!fluentMethods) { _%>set<%= mapsIdAssoc.relationshipNameCapitalized %> <%_ } else { _%><%= otherEntityName %><%_ } _%>);
     <%_ } _%>
-        <%= returnPrefix %> <%= entityInstance %>Repository.save(<%= persistInstance %>);
+        <%= returnPrefix %> <%= entityInstance %>Repository.saveAndFlush(<%= persistInstance %>);
```
Comment (mshima) [inline generators/entity-server/templates/partials/save_template.ejs:57]:
```suggestion
        <%= returnPrefix %> <%= entityInstance %>Repository.save(<%= persistInstance %>);
```

**Review by mshima — COMMENTED** [review]
Only sql provides saveAndFlush operations

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 21]
```diff
@@ -73,10 +84,25 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
 
     @Override
     public Flux<<%= persistClass %>> search(String query<% if (!paginationNo) { %>, Pageable pageable<% } %>) {
-        NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
   <%_ if (!paginationNo) { _%>
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:87]:
```suggestion
        NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
  <%_ if (!paginationNo) { _%>
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 22]
```diff
@@ -73,10 +84,25 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
+        NativeSearchQueryBuilder queryBuilder = new NativeSearchQueryBuilder().withQuery(queryStringQuery(query))
+            .withPageable(PageRequest.of(pageable.getPageNumber(), pageable.getPageSize()));
+
+        builders.stream().forEach(builder -> {
+            queryBuilder.withSort(builder);
+        });
+
+        NativeSearchQuery nativeSearchQuery = queryBuilder.build();
+  <%_ } else { _%>
+        NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:99]:
```suggestion
        nativeSearchQuery.setPageable(pageable);
```

**Review by mshima — COMMENTED** [review]
Reapply https://github.com/jhipster/generator-jhipster/commit/81eaa12358cbbd55867d93f30591b535e9c6fb7b

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 23]
```diff
@@ -87,39 +103,64 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
     }
   <%_ } _%>
+
+      @Override
+      @Async
+      @Transactional(propagation =  Propagation.REQUIRES_NEW, readOnly = true, isolation = Isolation.READ_COMMITTED)
+      public void index(<%= persistClass %> entity) {
+         repository.findOne<% if (implementsEagerLoadApis ) { %>WithEagerRelationships<% } %>(entity.getId()).ifPresent(elasticsearchTemplate::save);
+      }
+
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:164]:
```suggestion
    @Override
    @Async
    @Transactional(propagation =  Propagation.REQUIRES_NEW, readOnly = true, isolation = Isolation.READ_COMMITTED)
    public void index(<%= persistClass %> entity) {
        repository.find<% if (implementsEagerLoadApis ) { %>OneWithEagerRelationships<% } else { %>ById<% } %>(entity.getId()).ifPresent(elasticsearchTemplate::save);
    }

```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 24]
```diff
@@ -76,7 +87,12 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
         nativeSearchQuery.setPageable(pageable);
-  <%_ }_%>
+  <%_ } _%>
+        return search(nativeSearchQuery);
+    }
+
+    public Flux<<%= entityClass %>> search(Query query) {
+
         return reactiveElasticsearchTemplate
             .search(nativeSearchQuery, <%= persistClass %>.class)
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:97]:
```suggestion
            .search(query, <%= persistClass %>.class)
```

**Comment by ko5tik** [comment]
Hi All,    is it going to be merged anytime soon?  Do I need to make any changes?

**Comment by DanielFran** [comment]
@ko5tik Can you fix conflicts and revert changes in package-lock.json file?

**Comment by mshima** [comment]
Still missing the reactive part.

**Comment by ko5tik** [comment]
Rebased this branch onto current  development branch.   Somehow there are still conflicts

**Comment by OmarHawk** [comment]
> Rebased this branch onto current development branch. Somehow there are still conflicts

Also it still shows nearly 600 commits being part of the pull request. I guess, you should perform an actual rebase or just start with a clean branch based off current main branch. Would, by the way, also be interested in getting this merged. :-)

**Comment by ko5tik** [comment]
Now running into issue with integration test  - since indexing is now asynchronous,   current  integration test for createEntity()  and putNewEntity() are failiung -     as entities are not indexed yet.  Not sure how to wait for this properly

**Comment by ko5tik** [comment]
This pull request is outdated,    Changes were cherry picked into new branch  and new pull erquest created

https://github.com/jhipster/generator-jhipster/pull/18741

**Pull request closed without merging**